src/app/browser/qgsinbuiltdataitemproviders.cpp

```diff
@@ -612,3 +612,52 @@ bool QgsProjectItemGuiProvider::handleDoubleClick( QgsDataItem *item, QgsDataIte
     return false;
   }
 }
+
+//
+// QgsBookmarkItemGuiProvider
+//
+
+QString QgsBookmarkItemGuiProvider::name()
+{
+  return QStringLiteral( "bookmark_items" );
+}
+
+bool QgsBookmarkItemGuiProvider::handleDoubleClick( QgsDataItem *item, QgsDataItemGuiContext )
+{
+  if ( !item || item->type() != QgsDataItem::Bookmark )
+    return false;
+
+  if ( QgsBookmarkItem *bookmarkItem = qobject_cast<QgsBookmarkItem *>( item ) )
+  {
+    QgsReferencedRectangle rect = bookmarkItem->bookmark().extent();
+    QgsRectangle canvasExtent = rect;
+    if ( rect.crs() != QgisApp::instance()->mapCanvas()->mapSettings().destinationCrs() )
+    {
+      QgsCoordinateTransform ct( rect.crs(),
+                                 QgisApp::instance()->mapCanvas()->mapSettings().destinationCrs(), QgsProject::instance() );
+      try
+      {
+        canvasExtent = ct.transform( rect );
+      }
+      catch ( QgsCsException & )
+      {
+        QgisApp::instance()->messageBar()->pushWarning( tr( "Zoom to Bookmark" ), tr( "Could not reproject bookmark extent to project CRS." ) );
```

Comment:
Strictly speaking you should get the message bar from the context, but it's all the same in app